### PR TITLE
feat(web_console): job

### PR DESCRIPTION
**Pull request by nervouself**

### CODE REVIEW

File: web_console/api/job.js [hunk 1]
```diff
@@ -13,54 +21,200 @@ const namespace = config.NAMESPACE;
...
-  ctx.body = { data: pods };
+router.get('/api/v1/job/:k8s_name/pods', SessionMiddleware, async (ctx) => {
+  const { k8s_name } = ctx.params;
+  const { pods } = await k8s.getFLAppPods(namespace, k8s_name);
+  ctx.body = { data: pods.items };
 });
 
-router.get('/api/v1/job/pod/:name/:container', SessionMiddleware, async (ctx) => {
-  const { name, container } = ctx.params;
+router.get('/api/v1/job/:k8s_name/logs/:start_time', SessionMiddleware, async (ctx) => {
```
Comment (marswong) [inline web_console/api/job.js:66]: `/api/v1/job/:k8s_name/log?start_time=xxx` would be better, 'caz `es` query DSL is quite complicated

Comment (marswong) [reply web_console/api/job.js:66]: for example: we might wanna indicate `end_time` someday

File: web_console/api/job.js [hunk 2]
```diff
@@ -13,54 +21,200 @@ const namespace = config.NAMESPACE;
...
   const base = k8s.getBaseUrl();
-  const { id } = await k8s.getWebshellSession(namespace, name, container);
+  const { id } = await k8s.getWebshellSession(namespace, pod_name, container);
   ctx.body = { data: { id, base } };
 });
 
-router.get('/api/v1/job/pod/:name/logs/:time', SessionMiddleware, async (ctx) => {
-  const { name, time } = ctx.params;
-  const logs = await es.queryLog('filebeat-*', '', name, time, Date.now());
+router.get('/api/v1/job/pod/:pod_name/logs/:start_time', SessionMiddleware, async (ctx) => {
```
Comment (marswong) [inline web_console/api/job.js:79]: the same issue

File: web_console/api/job.js [hunk 3]
```diff
@@ -13,54 +21,200 @@ const namespace = config.NAMESPACE;
...
+    ctx.status = 422;
+    ctx.body = {
+      error: 'Job already exists',
     };
     return;
   }
-  const res = await k8s.createFLApp(namespace, fl_app);
-  ctx.body = { data: res };
+
+  const clientTicket = await Ticket.findOne({
```
Comment (marswong) [inline web_console/api/job.js:124]: `Ticket` takes priority over `Job`, return error if ticket not exits, no need for job validation

File: web_console/api/job.js [hunk 4]
```diff
@@ -13,54 +21,200 @@ const namespace = config.NAMESPACE;
...
+    ctx.body = {
+      error: 'client_params validation failed',
+    };
+    return;
+  }
+
+  // TODO: server validate
+  // TODO: server create job
+
+  const clientFed = await Federation.findByPk(clientTicket.federation_id);
```
Comment (marswong) [inline web_console/api/job.js:162]: why validate `Federation` here

File: web_console/libs/es.js [hunk 5]
```diff
@@ -7,64 +7,71 @@ const ky = require('ky-universal');
...
 
-    async queryLog(index, keyword, pod_name, start_time, end_time) {
-        let query_body = {
-            "sort": [
-                {
-                    "@timestamp": {
-                        "order": "asc",
-                        "unmapped_type": "boolean"
-                    }
+  async queryLog(index, keyword, pod_name, start_time, end_time, match_phrase) {
```
Comment (marswong) [inline web_console/libs/es.js:21]: only support `index` and `options/dsl` arguments would be better, and implement a utility function for `es` dsl generation

File: web_console/models/job.js [hunk 6]
```diff
@@ -42,6 +34,18 @@ module.exports = (sequelize, DataTypes) => {
       allowNull: true,
       default: null,
     },
+    user_id: {
```
Comment (marswong) [inline web_console/models/job.js:37]: no need. `job` would be created by `federation`

File: web_console/api/job.js [hunk 7]
```diff
@@ -13,54 +21,200 @@ const namespace = config.NAMESPACE;
...
 });
 
-router.post('/api/v1/job/create', SessionMiddleware, async (ctx) => {
-  const { fl_app } = ctx.request.body;
-  if (!fl_app) {
+// TODO: gRPC
+router.post('/api/v1/job', SessionMiddleware, async (ctx) => {
+  const { name, job_type, client_ticket_name, server_ticket_name } = ctx.request.body;
+
+  const [client_params_pass, client_params] = checkParseJson(ctx.request.body.client_params);
```
Comment (marswong) [inline web_console/api/job.js:89]: only use native `JSON.parse` and `JSON.stringify` for json validation

Comment (marswong) [reply web_console/api/job.js:89]: use `JSON object` type for `client_params` and `server_params` would be better

File: web_console/api/job.js [hunk 8]
```diff
@@ -13,54 +21,200 @@ const namespace = config.NAMESPACE;
...
+  if (!clientFed) {
+    ctx.status = 422;
+    ctx.body = {
+      error: 'Federation does not exist',
+    };
+    return;
+  }
+  const clientYaml = clientGenerateYaml(clientFed, job, clientTicket, serverTicket);
+
+  const res = await k8s.createFLApp(namespace, clientYaml);
```
Comment (marswong) [inline web_console/api/job.js:172]: @piiswrong client first or server first?

Comment (piiswrong) [reply web_console/api/job.js:172]: 应该是client收到server的response（success）以后再create的

File: web_console/api/job.js [hunk 9]
```diff
@@ -13,54 +21,200 @@ const namespace = config.NAMESPACE;
...
+    },
+  });
+  if (!clientTicket) {
+    ctx.status = 422;
+    ctx.body = {
+      error: 'client_ticket does not exist',
+    };
+    return;
+  }
+  const serverTicket = await Ticket.findOne({
```
Comment (marswong) [inline web_console/api/job.js:136]: validate by server

**Comment by marswong** [comment]
cc @piiswrong

**Pull request merged**